handler/incoming_instance_handler.go

```diff
@@ -108,25 +110,56 @@ func (hdlr *InstanceEventHandler) insertDimensions(ctx context.Context, instance
 
 	cache := make(map[string]string)
 
+	var wg sync.WaitGroup
+	problem := make(chan error, len(dimensions))
+	count := 0
+
 	for _, dimension := range dimensions {
+		wg.Add(1)
+		count++
+		go func(d *model.Dimension) {
+			defer wg.Done()
+
+			dbDimension, err := hdlr.Store.InsertDimension(ctx, cache, instance.DbModel().InstanceID, d.DbModel())
+			if err != nil {
+				err = errors.Wrap(err, "error while attempting to insert dimension")
+				log.Event(ctx, err.Error(), log.Error(err), log.Data{"instance_id": instance.DbModel().InstanceID, "dimension_id": dbDimension.DimensionID})
+				problem <- err
+				return
+			}
 
-		dbDimension, err := hdlr.Store.InsertDimension(ctx, cache, instance.DbModel().InstanceID, dimension.DbModel())
-		if err != nil {
-			return errors.Wrap(err, "error while attempting to insert dimension")
-		}
+			if err = hdlr.DatasetAPICli.PutDimensionNodeID(ctx, instance.DbModel().InstanceID, dimension); err != nil {
+				err = errors.Wrap(err, "DatasetAPICli.PutDimensionNodeID returned an error")
+				log.Event(ctx, err.Error(), log.Error(err), log.Data{"instance_id": instance.DbModel().InstanceID, "dimension_id": dbDimension.DimensionID})
+				problem <- err
+				return
+			}
 
-		if err = hdlr.DatasetAPICli.PutDimensionNodeID(ctx, instance.DbModel().InstanceID, dimension); err != nil {
-			return errors.Wrap(err, "DatasetAPICli.PutDimensionNodeID returned an error")
-		}
+			// todo: remove this temp hack once the time codelist / input data has been fixed.
+			if dbDimension.DimensionID != "time" {
+				if err = hdlr.Store.CreateCodeRelationship(context.Background(), instance.DbModel().InstanceID, dimension.CodeListID(), dbDimension.Option); err != nil {
+					err = errors.Wrap(err, "error attempting to create relationship to code")
+					log.Event(ctx, err.Error(), log.Error(err), log.Data{"instance_id": instance.DbModel().InstanceID, "dimension_id": dbDimension.DimensionID})
+					problem <- err
+					return
+				}
+			}
```

Comment:
Also I think it might be easier to follow if this were a separate function as this is technically its own _unit of work_ and there is a lot of code in this function already. Appreciate this is personal opinion but I find longer methods harder to follow as the scope of what its doing is usually much wider.

Reply:
I have extracted the function as suggested - I agree its much nicer having it separate 👍